### PR TITLE
Use helper methods to model resources

### PR DESCRIPTION
**Pull request by davidfowl**
- Remove the low level annotation usage of ContainerImageAnnotation and use the existing helper methods (to make sure they work as designed).
- Added tests for error conditions with helper methods
 ###### Microsoft Reviewers: [Open in CodeFlow](https://microsoft.github.io/open-pr/?codeflow=https://github.com/dotnet/aspire/pull/2777)

### CODE REVIEW

File: src/Aspire.Hosting/Extensions/ContainerResourceBuilderExtensions.cs [hunk 1]
```diff
@@ -160,9 +164,13 @@ public static IResourceBuilder<ContainerResource> AddContainer(this IDistributed
...
-        var containerImageAnnotation = builder.Resource.Annotations.OfType<ContainerImageAnnotation>().Single();
-        containerImageAnnotation.SHA256 = sha256;
-        return builder;
+        if (builder.Resource.Annotations.OfType<ContainerImageAnnotation>().LastOrDefault() is { } existingImageAnnotation)
+        {
+            existingImageAnnotation.SHA256 = sha256;
+            return builder;
+        }
+
+        throw new InvalidOperationException($"The resource '{builder.Resource.Name}' does not have a container image specified. Use WithImage to specify the container image and tag.");
```
Comment (eerhardt) [inline src/Aspire.Hosting/Extensions/ContainerResourceBuilderExtensions.cs:173]: (nit) maybe make a private helper for this exception so the message doesn't need to be duplicated (and doesn't get out of sync)?

File: src/Aspire.Hosting/Kafka/KafkaBuilderExtensions.cs [hunk 2]
```diff
@@ -24,9 +24,8 @@ public static IResourceBuilder<KafkaServerResource> AddKafka(this IDistributedAp
         var kafka = new KafkaServerResource(name);
         return builder.AddResource(kafka)
             .WithEndpoint(containerPort: KafkaBrokerPort, hostPort: port, name: KafkaServerResource.PrimaryEndpointName)
-            .WithAnnotation(new ContainerImageAnnotation { Image = "confluentinc/confluent-local", Tag = "7.6.0" })
-            .WithEnvironment(context => ConfigureKafkaContainer(context, kafka))
-            .PublishAsContainer();
```
Comment (eerhardt) [inline src/Aspire.Hosting/Kafka/KafkaBuilderExtensions.cs:29]: How does the manifest get written now if we don't call `PublishAsContainer()` by default?

**Review by eerhardt — APPROVED** [review]

File: src/Aspire.Hosting/Kafka/KafkaBuilderExtensions.cs (hunk 2 above)
Comment (davidfowl) [reply src/Aspire.Hosting/Kafka/KafkaBuilderExtensions.cs:29]: This is default for container resoures.

**Pull request merged**